labman/db/tests/test_study.py

```diff
@@ -82,6 +83,107 @@ def test_samples(self):
+                           'SKM4', 'SKM5', 'SKM6',
+                           'SKM7', 'SKM8', 'SKM9']
+            self.assertEqual(s.samples(), exp_samples)
+            exp_samples = ['SKB1', 'SKB2', 'SKB3',
+                           'SKB4', 'SKB5', 'SKB6',
+                           'SKB7', 'SKB8', 'SKB9']
+            self.assertEqual(s.samples(limit=9), exp_samples)
+            exp_samples = ['SKB1', 'SKB2', 'SKB3',
+                           'SKB4', 'SKB5', 'SKB6',
+                           'SKB7', 'SKB8', 'SKB9']
+            self.assertEqual(s.samples('SKB'), exp_samples)
+            exp_samples = ['SKB1', 'SKB2', 'SKB3']
+            self.assertEqual(s.samples('SKB', limit=3), exp_samples)
+            exp_samples = ['SKM1', 'SKM2', 'SKM3',
+                           'SKM4', 'SKM5', 'SKM6',
+                           'SKM7', 'SKM8', 'SKM9']
+            self.assertEqual(s.samples('SKM'), exp_samples)
+            exp_samples = ['SKM1', 'SKM2', 'SKM3',
+                           'SKM4', 'SKM5', 'SKM6',
+                           'SKM7', 'SKM8', 'SKM9']
+            # case insensitive
+            self.assertEqual(s.samples('Skm'), exp_samples)
+            self.assertEqual(s.samples('64'), [])
+
+            TRN.add(sql, [None])
+
+    def test_specimen_id_column(self):
+        s = Study(1)
+        self.assertIsNone(s.specimen_id_column)
+
+    def test_translate_ids_with_sample_id(self):
+        # Tests sample_id_to_specimen_id and specimen_id_to_sample_id when
+        # there is no specimen identifier set for the study
+        s = Study(1)
+
+        obs = s.sample_id_to_specimen_id('1.SKM4.640180')
+        self.assertEqual(obs, '1.SKM4.640180')
+
+        # doesn't even need to be a valid sample id
+        obs = s.sample_id_to_specimen_id('SKM3')
+        self.assertEqual(obs, 'SKM3')
+
+        with self.assertRaisesRegex(ValueError, 'Could not find "SKM4"'):
+            s.specimen_id_to_sample_id('SKM4')
+
+        with self.assertRaisesRegex(ValueError, 'Could not find "SSSS"'):
+            s.specimen_id_to_sample_id('SSSS')
+
+    def test_translate_ids_with_specimen_id(self):
+        s = Study(1)
+        # HACK: the Study object in labman can't modify specimen_id_column
+        # hence we do this directly in SQL, if a test fails the transaction
+        # will rollback, otherwise we reset the column to NULL.
+        sql = """UPDATE qiita.study
+                 SET specimen_id_column = %s
+                 WHERE study_id = 1"""
+        with sql_connection.TRN as TRN:
+            TRN.add(sql, ['anonymized_name'])
+
+            obs = s.sample_id_to_specimen_id('1.SKM4.640180')
+            self.assertEqual(obs, 'SKM4')
+
+            obs = s.specimen_id_to_sample_id('SKM4')
+            self.assertEqual(obs, '1.SKM4.640180')
+
+            # should be an exact match
+            with self.assertRaisesRegex(ValueError,
+                                        'Could not find \"1\.skm4\.640180\"'):
```

Comment:
Just to check, is case insensitivity a safe property of `sample_name`?

Reply:
Yes I think its fair to require exact case matching.